### PR TITLE
feat: support Kafka messages with nullable or empty partition key

### PR DESCRIPTION
**Pull request by sergioamribeiro**
# Description

This merge request adds protection for the kafka messages that arrive without a value for the partition key (null or empty). The Retry will receive the messages and put them in the same queue and follow that order to process

Fixes # ([issue](https://github.com/Farfetch/kafka-flow-retry-extensions/issues/10))

## How Has This Been Tested?

We've made some unit and integration tests (see `EmptyPartitionKeyRetryDurableTests` class)

## Checklist

-   [X] My code follows the style guidelines of this project
-   [X] I have performed a self-review of my own code
-   [X] I have added tests to cover my changes
-   [ ] I have made corresponding changes to the documentation

### Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

**Review by brunohfgomes — APPROVED** [review]

**Review by carlosgoias — DISMISSED** [review]

**Review by Daniel-C-Dias — DISMISSED** [review]

File: src/KafkaFlow.Retry/Durable/Repository/RetryDurableQueueRepository.cs [hunk 1]
```diff
@@ -257,6 +243,28 @@ private async Task<AddIfQueueExistsResult> AddIfQueueExistsAsync(IMessageContext
             }
         }
 
+        private RetryQueueItemMessage CreateRetryQueueItemMessage(IMessageContext context)
+        {
+            var partitionKey = default(byte[]);
```
Comment (fernando-a-marins) [inline src/KafkaFlow.Retry/Durable/Repository/RetryDurableQueueRepository.cs:248]: If we assume here the partitionKey as null (default(byte[])), we will get an SQL exception in the table `ItemMessage`, because the partition key field is not nullable. Furthermore, the kafka flow does not support produce a message with the null partition key as far as I know. PLease check the produce method.

File: src/KafkaFlow.Retry.IntegrationTests/EmptyPartitionKeyRetryDurableTests.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    using KafkaFlow.Retry.IntegrationTests.Core.Messages;
+    using KafkaFlow.Retry.IntegrationTests.Core.Producers;
+    using KafkaFlow.Retry.IntegrationTests.Core.Storages;
+    using KafkaFlow.Retry.IntegrationTests.Core.Storages.Assertion;
+    using KafkaFlow.Retry.IntegrationTests.Core.Storages.Repositories;
+    using Microsoft.Extensions.DependencyInjection;
+    using Xunit;
+
+    [Collection("BootstrapperHostCollection")]
+    public class EmptyPartitionKeyRetryDurableTests
```
Comment (fernando-a-marins) [inline src/KafkaFlow.Retry.IntegrationTests/EmptyPartitionKeyRetryDurableTests.cs:16]: It's missing tests with the null partition key. But please check my other comment.

**Review by fernando-a-marins — CHANGES_REQUESTED** [review]
It's not ready to support Null partition keys

File: src/KafkaFlow.Retry.IntegrationTests/EmptyPartitionKeyRetryDurableTests.cs (hunk 2 above)
Comment (spookylsm) [reply src/KafkaFlow.Retry.IntegrationTests/EmptyPartitionKeyRetryDurableTests.cs:16]: we can't test with a null partition because the Kafka flow doesn't support it.

File: src/KafkaFlow.Retry.IntegrationTests/Core/Storages/Assertion/RetryDurableGuaranteeOrderedConsumptionPhysicalStorageAssert.cs [hunk 3]
```diff
@@ -18,6 +18,29 @@ public RetryDurableGuaranteeOrderedConsumptionPhysicalStorageAssert(IRepositoryP
...
+
+            Assert.True(retryQueue.Id != Guid.Empty, "Retry Durable Creation Get Retry Queue cannot be asserted.");
+
+            var retryQueueItems = await this
+                .repositoryProvider
+                .GetRepositoryOfType(repositoryType)
+                .GetRetryQueueItemsAsync(retryQueue.Id, rqi => rqi.Count(item => item.Status == RetryQueueItemStatus.InRetry) != retryCount)
+                .ConfigureAwait(false);
+
+            Assert.True(retryQueueItems != null, "Retry Durable Creation Get Retry Queue Item Message cannot be asserted.");
```
[the remark this answers is not in the document]
Comment (spookylsm) [reply src/KafkaFlow.Retry.IntegrationTests/Core/Storages/Assertion/RetryDurableGuaranteeOrderedConsumptionPhysicalStorageAssert.cs:37]: if we use `Assert.NotNull` we lost the message, but if it's not important we can change.

File: src/KafkaFlow.Retry.IntegrationTests/Core/Storages/Assertion/RetryDurableLatestConsumptionPhysicalStorageAssert.cs [hunk 4]
```diff
@@ -18,6 +19,37 @@ public RetryDurableLatestConsumptionPhysicalStorageAssert(IRepositoryProvider re
...
+                {
+                    return rqi.Count(item => item.Status == RetryQueueItemStatus.Waiting) != retryCount;
+                }).ConfigureAwait(false);
+
+            var lastRetryItem = retryQueueItems.OrderBy(x => x.Sort).Last();
+            var numberOrRetryItems = retryQueueItems.Count();
+            var maxSortValue = retryQueueItems.Max(i => i.Sort);
+            var cancelledRetryItems = retryQueueItems.Except(new List<RetryQueueItem> { lastRetryItem });
+
+            Assert.True(retryQueueItems != null, "Retry Durable Retrying Get Retry Queue Item Message cannot be asserted.");
```
[the remark this answers is not in the document]
Comment (spookylsm) [reply src/KafkaFlow.Retry.IntegrationTests/Core/Storages/Assertion/RetryDurableLatestConsumptionPhysicalStorageAssert.cs:46]: https://github.com/Farfetch/kafka-flow-retry-extensions/pull/87#discussion_r883745334

File: src/KafkaFlow.Retry/Durable/Repository/RetryDurableQueueRepository.cs [hunk 5]
```diff
@@ -279,6 +287,21 @@ private RetryDurableException GetCheckQueueException(string message, QueueNewest
             return kafkaException;
         }
 
+        private string GetQueueGroupKey(object messageKey)
+        {
+            string messageKeyAsString;
```
[the remark this answers is not in the document]
Comment (spookylsm) [reply src/KafkaFlow.Retry/Durable/Repository/RetryDurableQueueRepository.cs]: please take a look again

File: src/KafkaFlow.Retry/Durable/Repository/RetryDurableQueueRepository.cs (hunk 1 above)
Comment (luispfgarces) [reply src/KafkaFlow.Retry/Durable/Repository/RetryDurableQueueRepository.cs:248]: Maybe a `new byte[0]` solves the issue?

Let me also suggest to use a ternary condition here:

```csharp
var partitionKey = context.Message.Key is object ? Encoding.ASCII.GetBytes(JsonConvert.SerializeObject(context.Message.Key)) : new byte[0];
```

Because creating an empty byte array always incurs a memory cost. I know, almost negligible, but since we are here 🙂

File: src/KafkaFlow.Retry.IntegrationTests/Core/Storages/Assertion/RetryDurableGuaranteeOrderedConsumptionPhysicalStorageAssert.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (fernando-a-marins) [reply src/KafkaFlow.Retry.IntegrationTests/Core/Storages/Assertion/RetryDurableGuaranteeOrderedConsumptionPhysicalStorageAssert.cs:37]: ok. So it's better the Assert.True then

File: src/KafkaFlow.Retry/Durable/Repository/RetryDurableQueueRepository.cs (hunk 1 above)
Comment (fernando-a-marins) [reply src/KafkaFlow.Retry/Durable/Repository/RetryDurableQueueRepository.cs:248]: Yes, I agree with you @luispfgarces

File: src/KafkaFlow.Retry.IntegrationTests/EmptyPartitionKeyRetryDurableTests.cs (hunk 2 above)
Comment (fernando-a-marins) [reply src/KafkaFlow.Retry.IntegrationTests/EmptyPartitionKeyRetryDurableTests.cs:16]: You can test it if the test does not start with the production of a Kafka message but with a consumption. If we want to support null partition keys we should have a test for it.
